_posts/2018-10-08-http-part5.markdown

```diff
@@ -0,0 +1,305 @@
+
+After a first request from a connection, our server will go through second iteration of the loop and will block on `Request.fromRaw(input)`, (6) which will block on `input.readLine()` of `BufferedReader`, which internally will block on a socket `read()`. This loop ends if either server or client closes the connection (7) or will send `Connection: close` header (8).
+
+#### Detecting connections closed by client
+Detecting closed connections by our server is easy, we can check `socket.isClosed`. The tricky part for me was to check if a connection was closed by the client. When the client closes a connection, it sends `FIN,ACK` segment and we send the `ACK` back.  At first, I thought that we can just check `socket.isClosed` too, but I was wrong - this flag only checks if *we* closed the socket.
+[It seems](https://stackoverflow.com/a/10241044/1189159) that there is no information about `FIN,ACK` event from the client in Socket API.
+The only indication of it is to try to either use `read()` or `readLine()` on `BufferedReader`.
+
+With that in mind, it is hard to do the following logic - block until new data arrives, then parse the request, if there is no data, close a connection. We cannot use `readLine()` before `Request.fromRaw(input)` because it will consume one line of our `BufferedReader`. We could change signature `fromRaw` to accept the first line as `String` and the rest of a stream as `BufferedReader` but that’s just ugly.
+That’s why for now we can keep things as they were. The beginning of our parsing logic protects us from empty requests anyways.
+``` kotlin
+val requestLine = input.readLine() ?: throw EmptyRequestException()
+```
+If `readLine()` returns null, we can just catch it (7) and close the connection. It might not be the best, it may smell like using exceptions for control flow, but it works. 
+We will try to tackle this problem in the future maybe by using some sort of “one line/character caching BufferedReader”.
+
+## Tests
+The main happy path to test is to check whether we can make several requests on one connection.
+``` kotlin
+private val client = HttpUrlConnectionClient(connectionTimeout = 50, socketTimeout = 100)
+
+@Test
+fun `should reuse connection to make requests`() {
+    // given
+    server = Server(handler = {
+        Response(status = Status.OK, headers = HttpHeaders("content-length" to "0"))
+    })
+    server.start(8090)
+
+    // when
+    repeat(times = 3) {
+        client.exchange(url = "http://localhost:8090", request = Request(path = "/", method = RequestMethod.GET))
+    }
+
+    // then
+    assertThat(server.connectionsEstablished()).isEqualTo(1)
+}
+```
+To test that, we have to expose `connectionsEstablished()` method that returns the count of connections established to the server. This most likely will be transformed in some kind of `ServerStats` object, because we will want to have metrics, but let’s keep it simple for now.
```

Comment:
To test that we have to expose `connectionsEstablished()`, method that returns the count of...